File: LeetCode/Medium/0316-remove-duplicate-letters/0316-remove-duplicate-letters.java
class Solution {
    public String removeDuplicateLetters(String s) {
        int[] counter = new int[26];
        boolean[] used = new boolean[26];

        for (char c : s.toCharArray()) {
            counter[c - 'a']++;
        }

        StringBuilder sb = new StringBuilder();

        for (char c : s.toCharArray()) {
            int idx = c - 'a';
            counter[idx]--;

            if (used[idx]) continue;

            while (sb.length() > 0) {
                int last = sb.length() - 1;

                if (sb.charAt(last) > c && counter[sb.charAt(last) - 'a'] > 0) {
                    used[sb.charAt(last) - 'a'] = false;
                    sb.deleteCharAt(last);
                } else {
                    break;
                }
            }

            sb.append(c);
            used[idx] = true;
        }

        return sb.toString();
    }
}

File: LeetCode/Medium/0316-remove-duplicate-letters/0316-remove-duplicate-letters.kt
class Solution {
    fun removeDuplicateLetters(s: String): String {
        val counter = IntArray(26)
        val used = BooleanArray(26)

        for (c in s) {
            counter[c - 'a']++
        }

        val sb = StringBuilder()

        for (c in s) {
            val idx = c - 'a'
            counter[idx]--

            if (used[idx]) continue

            while (sb.isNotEmpty()) {
                val last = sb.last()

                if (last > c && counter[last - 'a'] > 0) {
                    used[last - 'a'] = false
                    sb.deleteAt(sb.lastIndex)
                } else {
                    break
                }
            }

            sb.append(c)
            used[idx] = true
        }

        return sb.toString()
    }
}